File: packages/node/src/routing/handlers/chat-request.ts
/**
 * Handlers for Telegram Chat Join Requests and Chat Boost updates.
 *
 * @packageDocumentation
 */

import type { Update } from "../../kernel/update.js";
import { CallbackContext } from "../../kernel/context.js";
import { BaseHandler, HandlerCallback } from "./base.js";

/**
 * Handler for user requests to join private chats or channels (`chat_join_request` updates).
 *
 * @typeParam C - Type of the callback context.
 * @typeParam R - Return value type.
 *
 * @example
 * ```ts
 * const handler = new ChatJoinRequestHandler(async (update, context) => {
 *   await context.bot.approveChatJoinRequest(
 *     update.chat_join_request!.chat.id,
 *     update.chat_join_request!.from.id,
 *   );
 * });
 * ```
 */
export class ChatJoinRequestHandler<
  C extends CallbackContext = CallbackContext,
  R = unknown,
> extends BaseHandler<C, R> {
  /**
   * Checks whether the incoming update contains a `chat_join_request`.
   *
   * @param update - The update to test.
   * @returns `true` if update has `chat_join_request`.
   */
  async checkUpdate(update: Update): Promise<boolean> {
    return Boolean(update.chat_join_request);
  }
}

/**
 * Handler for chat boost additions and removals (`chat_boost` and `removed_chat_boost` updates).
 *
 * @typeParam C - Type of the callback context.
 * @typeParam R - Return value type.
 *
 * @example
 * ```ts
 * const handler = new ChatBoostHandler(async (update, context) => {
 *   if (update.chat_boost) {
 *     console.log("Chat boosted by user:", update.chat_boost.boost.source);
 *   }
 * });
 * ```
 */
export class ChatBoostHandler<
  C extends CallbackContext = CallbackContext,
  R = unknown,
> extends BaseHandler<C, R> {
  /** Target only added or updated boosts (`chat_boost`) */
  public static readonly ADDED = 1;
  /** Target only removed boosts (`removed_chat_boost`) */
  public static readonly REMOVED = 2;
  /** Target any boost event */
  public static readonly ANY = 3;

  public readonly boostTypes: number;

  /**
   * Constructs a new {@link ChatBoostHandler}.
   *
   * @param callback - Function invoked when the update matches.
   * @param boostTypes - Filter mask (`ADDED`, `REMOVED`, or `ANY`).
   * @defaultValue `ChatBoostHandler.ANY`
   */
  constructor(callback: HandlerCallback<C, R>, boostTypes: number = ChatBoostHandler.ANY) {
    super(callback);
    this.boostTypes = boostTypes;
  }

  /**
   * Checks whether the incoming update matches the configured boost filter.
   *
   * @param update - The update to test.
   * @returns `true` if update matches.
   */
  async checkUpdate(update: Update): Promise<boolean> {
    if (this.boostTypes === ChatBoostHandler.ADDED) {
      return Boolean(update.chat_boost);
    }
    if (this.boostTypes === ChatBoostHandler.REMOVED) {
      return Boolean(update.removed_chat_boost);
    }
    return Boolean(update.chat_boost || update.removed_chat_boost);
  }
}

File: packages/python/src/telebot_py/routing/handlers/chat_request.py
"""Handlers for chat join requests and chat boost updates."""

from __future__ import annotations

from telebot_py.routing.handlers.base import BaseHandler, HandlerCallback
from telebot_py.types import Update


class ChatJoinRequestHandler(BaseHandler):
    """Handler for requests to join private chats or channels.

    Matches ``chat_join_request`` updates, mirroring the node
    ``ChatJoinRequestHandler``.
    """

    def check_update(self, update: object) -> bool:
        """Return whether the update carries a chat join request.

        Args:
            update: The incoming update.

        Returns:
            ``True`` when the update carries a ``chat_join_request`` payload.
        """
        return isinstance(update, Update) and update.chat_join_request is not None


class ChatBoostHandler(BaseHandler):
    """Handler for chat boost additions and removals.

    Matches ``chat_boost`` and/or ``removed_chat_boost`` updates depending on
    the :attr:`boost_types` filter mask, mirroring the node
    ``ChatBoostHandler`` constants.

    Attributes:
        boost_types: Filter mask selecting which boost events to handle
            (:attr:`ADDED`, :attr:`REMOVED`, or :attr:`ANY`).
    """

    #: Target only added or updated boosts (``chat_boost``).
    ADDED = 1
    #: Target only removed boosts (``removed_chat_boost``).
    REMOVED = 2
    #: Target any boost event.
    ANY = 3

    def __init__(
        self,
        callback: HandlerCallback,
        boost_types: int = ANY,
        *,
        block: bool = True,
    ) -> None:
        """Initialize the handler.

        Args:
            callback: Callable invoked when a matching boost update arrives.
            boost_types: Filter mask (`ADDED`, `REMOVED`, or `ANY`).
                Defaults to `ChatBoostHandler.ANY`.
            block: Whether dispatch awaits the callback before moving on.
        """
        super().__init__(callback, block=block)
        self.boost_types = boost_types

    def check_update(self, update: object) -> bool:
        """Return whether the update matches the configured boost filter.

        Args:
            update: The incoming update.

        Returns:
            ``True`` when the update carries a boost event allowed by the
            configured mask, ``False`` otherwise.
        """
        if not isinstance(update, Update):
            return False
        if self.boost_types == ChatBoostHandler.ADDED:
            return update.chat_boost is not None
        if self.boost_types == ChatBoostHandler.REMOVED:
            return update.removed_chat_boost is not None
        return update.chat_boost is not None or update.removed_chat_boost is not None
